src/NServiceBus.Core/Unicast/Messages/MessageMetadataRegistry.cs

```diff
@@ -48,12 +48,21 @@ public MessageMetadata GetMessageMetadata(string messageTypeIdentifier)
         {
             Guard.AgainstNullAndEmpty(nameof(messageTypeIdentifier), messageTypeIdentifier);
 
-            var messageType = Type.GetType(messageTypeIdentifier, false);
+            var messageType = GetType(messageTypeIdentifier);
 
             if (messageType == null)
             {
                 Logger.DebugFormat("Message type: '{0}' could not be determined by a 'Type.GetType', scanning known messages for a match", messageTypeIdentifier);
-                return messages.Values.FirstOrDefault(m => m.MessageType.FullName == messageTypeIdentifier);
+
+                foreach (var item in messages.Values)
+                {
+                    if (item.MessageType.FullName == messageTypeIdentifier)
```

Comment:
good catch, my example was wrong. It works when the other side specifies the fullname explicitly in the header

Like:

https://github.com/Particular/ServiceControl.Plugin.Nsb6.Heartbeat/blob/master/src/ServiceControl.Plugin.Nsb5.Heartbeat/ServiceControlBackend.cs#L39

and potentially here if using a serializer that can't infer types from the payload (json.net can using $type and xml can using the root node name)

http://docs.particular.net/samples/sqltransport/native-integration/?version=sqlserver_3#code-walk-through-sending-the-message

Side note: To fix the mismatching asm versions we should stop leaking platform specifics onto the wire like suggested here:

https://github.com/Particular/NServiceBus/issues/1610

sorry for the wild goose chase!